src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+consists of two steps:
+
+1.  Update the `shared_ptr` to the current node.
+2.  Update the `SHAMapNodeID`.
+
+The first step consists of several attempts to find the node in various places:
+
+1.  In the trie itself.
+2.  In the node cache.
+3.  In the database.
+
+If the node is found in one of the second-two places, then it is installed into
+the trie as part of the traversal process.
 
 ## Late-arriving Nodes ##
 
-As we noted earlier, SHAMaps (even immutable ones) may grow.  If a SHAMap
-is searching for a node and runs into an empty spot in the tree, then the
-SHAMap looks to see if the node exists but has not yet been made part of
-the map.  This operation is performed in the `SHAMap::fetchNodeExternalNT()`
-method.  The *NT* is this case stands for 'No Throw'.
+As we noted earlier, SHAMaps (even immutable ones) may grow.  If a SHAMap is
+searching for a node and runs into an empty spot in the trie, then the SHAMap
+looks to see if the node exists but has not yet been made part of the map.  This
+operation is performed in the `SHAMap::fetchNodeExternalNT()` method.  The *NT*
+is this case stands for 'No Throw'.
 
 The `fetchNodeExternalNT()` method goes through three phases:
 
  1. By calling `getCache()` we attempt to locate the missing node in the
-    TreeNodeCache.  The TreeNodeCache is a cache of immutable
-    SHAMapTreeNodes that are shared across all SHAMaps.
+    TreeNodeCache.  The TreeNodeCache is a cache of immutable SHAMapTreeNodes
+    that are shared across all SHAMaps.
 
-    Any SHAMapTreeNode that is immutable has a sequence number of zero.
-    When a mutable SHAMap is created then its SHAMapTreeNodes are given
-    non-zero sequence numbers.  So the `assert (ret->getSeq() == 0)`
-    simply confirms that the TreeNodeCache indeed gave us an immutable node.
+    Any SHAMapTreeNode that is immutable has a sequence number of zero
+    (sharable). When a mutable SHAMap is created then its SHAMapTreeNodes are
+    given non-zero sequence numbers (unsharable).  So the `assert (ret->getSeq()
+    == 0)` simply confirms that the TreeNodeCache indeed gave us an immutable
+    node.
 
  2. If the node is not in the TreeNodeCache, we attempt to locate the node
-    in the historic data stored by the data base.  The call to
-    to `fetch(hash)` does that work for us.
+    in the historic data stored by the data base.  The call to to `fetch(hash)`
+    does that work for us.
 
  3. Finally, if ledgerSeq_ is non-zero and we did't locate the node in the
     historic data, then we call a MissingNodeHandler.
 
     The non-zero ledgerSeq_ indicates that the SHAMap is a complete map that
-    belongs to a historic ledger with the given (non-zero) sequence number.
-    So, if all expected data is always present, the MissingNodeHandler should
-    never be executed.
+    belongs to a historic ledger with the given (non-zero, unsharable) sequence
+    number. So, if all expected data is always present, the MissingNodeHandler
```

Comment:
Might be worth adding a description on what the MissingNodeHandler is and does.

Reply:
Looks like this section is obsolete too, rewritting...